examples_tests/35.TAA/main.cpp

```diff
@@ -294,17 +289,40 @@ int main()
     TAAMaterial.setTexture(3u, velocityBuf);
     TAAMaterial.TextureLayer[3].SamplingParams = sparams;
     TAAMaterial.MaterialType = TAAMaterialType;
+    TAAMaterial.ZBuffer = video::ECFN_ALWAYS;
+    TAAMaterial.ZWriteEnable = false;
 
 	uint64_t lastFPSTime = 0;
 
+    core::matrix4SIMD prevVP, currVP;
 	while(!quit && device->run())
 	{
 		driver->beginScene(false, false, video::SColor(255,255,255,255) );
 
         const float clearVel[4]{ 0.f, 0.f, 0.f, 0.f };
-        const float clearColor[4]{ 0.f, 0.f, 1.f, 1.f };
+        const float clearColor[4]{ 0.9f, 0.9f, 0.9f, 1.f };
         const float clearDepth = 0.f;
 
+        //update prevVP and currVP before velocity pass
+        if (g_FrameNum==0u)
+        {
+            currVP = prevVP = driver->getTransform(video::EPTS_PROJ_VIEW);
+        }
+        else
+        {
+            prevVP = currVP;
+            currVP = driver->getTransform(video::EPTS_PROJ_VIEW);
+#if USE_JITTER
+            core::matrix4SIMD jitter;
+            float jitterOffset[3]{ 0.f,0.f,0.f };
+            memcpy(jitterOffset, &JITTER_OFFSETS[g_FrameNum%JITTER_OFFSET_CNT].X, 8);
+            jitter.setTranslation(jitterOffset);
```

Comment:
apply Jitter onto the camera proj matrix before it `render`s or `drawAll` is called, that will make all matrices consistent (all combos of proj matrix) also right now you're not rendering with the jitter!

Secondly you've hardcoded the `JITTER_OFFSETS` to have the window resolution inside, it would be far better to have them [0,1)^2 pixel footprint relative sobol or halton samples (or standard Vulkan MSAA pattern sequence) that are later expanded to viewport resolution (not render target resolution, but active viewport resolution).

Furthermore your offsets are miscalculated because NDC is in [-1,1]^2 so your jitters need to be in range [-1/RES_X,1/RES_X]*[-1/RES_Y,1/RES_Y], so approximately 2x too small and biased to one side

Lastly, the matrix concatenation is wrong, there will be a W divide at the end, and that will make the vanishing point at (0,0) NDC, when in-fact for a jittered view, the vanishing point should be at `jitter_offset`.
https://computergraphics.stackexchange.com/questions/6021/changing-the-projection-reference-point-in-open-gl
Which means that you also need to modify column 3 to add a `jitter` proportional factor for the viewspace Z component, so the vanishing line/point stays the same in world coordinates.

Reply:
@Crisspl PAge 11 http://twvideo01.ubm-us.net/o1/vault/gdc2016/Presentations/Pedersen_LasseJonFuglsang_TemporalReprojectionAntiAliasing.pdf

> use projection offset to shear frustum

You don't have the shear that maintains the same vanishing point

Reply:
Look here at `a`,`b`,`c`
https://github.com/playdeadgames/temporal/blob/master/Assets/Scripts/Extensions.cs#L74

Its a shear

Reply:
Also lots of impl details here
http://advances.realtimerendering.com/s2016/s16_Ke.pptx